tests/PHPStan/Analyser/nsrt/uppercase-string-trim.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace UppercaseStringTrim;
+
+use function PHPStan\Testing\assertType;
+
+class Foo
+{
+
+	/**
+	 * @param uppercase-string $uppercase
+	 */
+	public function doTrim(string $uppercase, string $string): void
+	{
+		assertType('lowercase-string|uppercase-string', trim($uppercase));
+		assertType('lowercase-string|uppercase-string', ltrim($uppercase));
+		assertType('lowercase-string|uppercase-string', rtrim($uppercase));
+		assertType('lowercase-string|uppercase-string', trim($uppercase, $string));
+		assertType('lowercase-string|uppercase-string', ltrim($uppercase, $string));
+		assertType('lowercase-string|uppercase-string', rtrim($uppercase, $string));
```

Comment:
The fact tests are green with `lowercase-string|uppercase-string` is weird.

I recommend to write `assertType('uppercase-string', ...);` and we'll need to debug this.

Reply:
I feel like the trim issue is because of `uppercaseString::isLowercase` is maybe and how conditional return type works.

I tried and the issue is solved with a real dynamicReturnTypeExtension, you should cherry-pick/copy this https://github.com/phpstan/phpstan-src/commit/22f0cab4b286a0e819056ff07dd3680ccba6e9f6

Maybe the same thing could be done for parse_str ? (and mb_parse_str, I just discovered)...
but I dunno if there is something similar to dynamicReturnTypeExtension for param-out.

Reply:
I see them now, and will copy -- thanks!

Reply:
@VincentLanglet Copied and added in ebd50a326 -- however, [tests/PHPStan/Analyser/data/param-out.php](https://github.com/phpstan/phpstan-src/pull/3613/files#diff-0718f319732d6bab708878e5f02d9cd20d0f7a135f2f6bb6a3cbc89c6547febe) began to fail, so I modified the test to update the assertion. Let me know if that should be reverted.

Reply:
Excellent -- resolving this conversation, please unresolve as you see fit.